[Tooling] Improvements on the AppLocalizedString linter

AliSoftware opened the pull request:
This is a follow-up on https://github.com/wordpress-mobile/WordPress-iOS/pull/18462 to address the remaining improvements I had in mind for the swift implementation back then:

## What was changed

 - Turn on optimizations when compiling the swift code (honestly doesn't change much, but 🤷 )
 - Remove debugging leftovers, fix typos in comments
 - Better strategy for parsing the `PBXObjects` based on their `isa`
 - Fix handling of `XCVersionGroup` and `PBXVariantGroup`
    - First by better understanding how to deal with the case of a `PBXBuildFile` pointing to such group was supposed to be handled in the first place…
    - … then by deciding to ignore those cases completely because they are only used for compiled resource "packages" like localized files (`*.strings`) and versioned files (`xcdatamodeld`), but we're only interested in source files (`*.swift`/`*.m`) anyway
 - Remove all force-unwraps and `fatalError`
    - Replace them with thrown errors
    - Provide better error messages and error reporting at top-level for any `.pbxproj` inconsistencies
    - More consistent reporting when encountering objects of known `isa` but with invalid data — as opposed to parse them as `UnknownPXBObject` at first and throw an unrelated error like `.orphanObject` only later
    - Protect against unknown values of some `PBXObject`'s fields, like an unknown `productType` or a `PBXBuildFile` pointing to a `productRef` (e.g. Swift Package dependency) instead of a `fileRef`

## To Test

 - Check that the project compiles in Xcode with a clean build
 - Check that the project compiles in Xcode with an incremental build
 - Check that the time taken by the Script Build phase in the build logs didn't jump to some crazy value and taking too long
 - Introduce a violation by replacing a random `AppLocalizedString` in code by `NSLocalizedString`, rebuild and validate the linter still detects it
 - Introduce an error in the `pbxproj` file to make it invalid, then check that the running the script from the terminal shows the expected error nicely. e.g.:
    - Replacing a `sourceTree = "<group>"` by `sourceTree = "wut"` to get a `DecodingError`
    - Changing a `PBXGroup` to have a `sourceTree = "<absolute>"` but removing its `path = …` value to get the `.incorrectAbsolutePath` error
    - Find the UUID of one of the parsed file, and remove it from the `children` list of its parent `PBXGroup` to get the `.orphanObject` error
    - Change an UUID randomly to get the `.objectNotFound` error

AliSoftware committed:
Enable optimizations when compiling the swift script

AliSoftware committed:
Remove leftover from debugging code

AliSoftware committed:
Fix typo in comment

AliSoftware committed:
Fix `resolveURL` for `PBXBuildFiles` pointing to groups

If a `PBXBuildFile` point to a `PBXVariantGroup` (e.g. `Localizable.strings`) or a `XCVersionGroup` (e.g. `*.xcdatamodeld`) that means that all the `children` files of those groups are going to be built

AliSoftware committed:
Catch `sourceTree=<absolute>` with `nil` path

AliSoftware committed:
Better args handling of shell wrapper script

To catch missing project path, and also allow extra arguments and pass them by to the swift script

Mostly useful when testing the script in Terminal (e.g. if we want to test the script on a single target) - because when called from the Xcode's Script Build Phases, we don't pass any argument to the wrapper script anyway

AliSoftware committed:
Stop resolving URLs of `PBXBuildFile` of groups

If a PBXBuildFile ends up pointing to a `PBXVariantGroup` or `XCVersionGroup` (as opposed to a `PBXFileReference`), this means to Xcode that all the `children` in that group will be built as part of the build phase.
For example, for a `PBXBuildFile` pointing to a `PBXVariantGroup` for `Localizable.strings`, Xcode will actually build all the `*.lproj/Localizable.strings` files that are `children` to that `PBXVariantGroup`.

In practice though, since we only care about linting `.string` and `.m` files and not `.strings` or `xcdatamodeld` and similar file-groupings, it's easier just not to bother with those and not even try to `resolveURL` on those since we'll skip them anyway when comes the call to `lint(…)`

AliSoftware committed:
Better implementation of `isa`-based decoding

AliSoftware committed:
Warn about and skip unhandled `sourceTree` values

Instead of crashing with `fatalError`

AliSoftware committed:
Stop handling `PBXVariantGroup`/`XCVersionGroup`

We don't need to handle those for our needs, as the only files we want to lint at `.swift`/`.m` source files, while `PBXVariantGroup` are for localized resources (e.g. `*.strings` inside `*.lproj` folders) and `XCVersionGroup` are for versioned resources (e.g. `*.xcdatamodel` files inside `*.xcdatamodeld` bundles), none of which we need to handle after all.

AliSoftware committed:
Use `throw` instead of `fatalError`

AliSoftware committed:
Catch top-level thrown errors & print them nicely

AliSoftware committed:
Allow parsing `PBXBuildFile` without a `fileRef`

Some `PBXBuildFile` are pointing to a `productRef` for example (e.g. SPM dependency) and while we don't need to parse and handle those, we don't want to have the `PropertyListDecoder` throw during parsing for missing that `fileRef` property for those.

AliSoftware committed:
Enforce decoding for objects of known `isa`

Before that change, if we found an object with a known `isa` value, we'd silently eat any error while parsing that known object.

Which could then lead to cryptic or misleading errors later down the line — e.g. an error about "failing to find the parent of an object" because we earlier silently failed to parsed that parent `PBXObject` as a typed object, instead of reporting the error at the point where the parsing of the `PBXObject` failed.

With that change if we find a known type but there's an error while trying to decode that known type, we'll now report it correctly.

AliSoftware committed:
Allow unknown productType for `PBXNativeTarget`

So that any future, unknown `productType` value won't make the parsing of the `pbxproj` crash, but instead parse the `PBXNativeTarget` normally and just record the unknown value as `String`.

AliSoftware committed:
Simpler implementation for unknown `productType`

AliSoftware committed:
Let Xcode fix indentation

AliSoftware commented on Scripts/BuildPhases/LintAppLocalizedStringsUsage.sh:
```diff
@@ -8,10 +8,14 @@ LINTER_EXEC="${LINTER_BUILD_DIR}/$(basename "${SCRIPT_SRC}" .swift)"
 if [ ! -x "${LINTER_EXEC}" ] || ! (shasum -c "${LINTER_EXEC}.shasum" >/dev/null 2>/dev/null); then
   echo "Pre-compiling linter script to ${LINTER_EXEC}..."
-  swiftc -sdk "$(xcrun --sdk macosx --show-sdk-path)" "${SCRIPT_SRC}" -o "${LINTER_EXEC}"
+  swiftc -O -sdk "$(xcrun --sdk macosx --show-sdk-path)" "${SCRIPT_SRC}" -o "${LINTER_EXEC}"
```
After a quick test, turning optimizations on apparently does not win us much execution time for it to be significant… but also doesn't hurt to keep… 🤷

AliSoftware commented on Scripts/BuildPhases/LintAppLocalizedStringsUsage.swift:
```diff
@@ -6,30 +6,30 @@ import Foundation
     let projectURL: URL // points to the "<projectDirectory>/<projectName>.xcodeproj/project.pbxproj" file
     private let pbxproj: PBXProjFile
-
+    
```
In the previous iteration, I used a mix of Xcode playgrounds to experiment with the initial concept… then used VSCode and the Swift LSP plugin to write the final script (and I guess VSCode does not indent empty lines at all).

When I worked on those improvements in this PR, I edited the `.swift` file directly in Xcode, then hit Ctrl-I to re-indent. That's when Xcode decided to add those indentations even on empty lines (and despite my Xcode Preferences telling to also trim whitespace-only lines… 🤷 )

I decided to keep those because it's most likely that anyone who might edit that `.swift` file again in the future will also use Xcode anyway.

mokagio replied:
Xcode and indenting empty lines with `Crtl I` even when you tell it not to... then removing those indentation if you move your cursor over them 😭 🤦 😡

AliSoftware commented on Scripts/BuildPhases/LintAppLocalizedStringsUsage.swift:
```diff
@@ -6,30 +6,30 @@ import Foundation
     private lazy var referrers: [ObjectUUID: ObjectUUID] = {
         var referrers: [ObjectUUID: ObjectUUID] = [:]
-        func recurseIfGroup(objectID: ObjectUUID, indent: String = "") {
+        func recurseIfGroup(objectID: ObjectUUID) {
```
Leftover from debugging during the previous iteration of that script

AliSoftware commented on Scripts/BuildPhases/LintAppLocalizedStringsUsage.swift:
```diff
@@ -39,54 +39,61 @@ class Xcodeproj {
-            let fileRef = try! self.pbxproj.object(id: buildFile.fileRef) as PBXGroup
-            return resolveURL(objectUUID: buildFile.fileRef, object: fileRef)
+    func resolveURL(to buildFile: PBXBuildFile) throws -> URL? {
+        if let fileRefID = buildFile.fileRef, let fileRefObject = try? self.pbxproj.object(id: fileRefID) as PBXFileReference {
+            return try resolveURL(objectUUID: fileRefID, object: fileRefObject)
+        } else {
```
`PBXBuildFile.fileRef` is now an optional (`ObjectUUID?`) to cover for the case of `PBXBuildFile` objects in the `.pbxproj` pointing to things other than a `fileRef`.

One of such cases are `PBXBuildFile` objects pointing to a `productRef` (e.g. a Swift Package dependency) instead of a `fileRef`.

We want to silently ignore the case of those type of `PBXBuildFiles` (as opposed to crashing during decoding of the plist, or trying to implement `struct PBXBuildFile` to parse them while we don't even have a need to handle those), and for those cases just make `resolveURL` return `nil`, hence the `if let fileRefID = buildFileRef` here now.

AliSoftware commented on Scripts/BuildPhases/LintAppLocalizedStringsUsage.swift:
```diff
@@ -39,54 +39,61 @@ class Xcodeproj {
+        if let fileRefID = buildFile.fileRef, let fileRefObject = try? self.pbxproj.object(id: fileRefID) as PBXFileReference {
+            return try resolveURL(objectUUID: fileRefID, object: fileRefObject)
+        } else {
+            // If the `PBXBuildFile` is pointing to `XCVersionGroup` (like `*.xcdatamodel`) and `PBXVariantGroup` (like `*.strings`)
+            // (instead of a `PBXFileReference`), then in practice each file in the group's `children` will be built by the Build Phase.
+            // In practice we can skip parsing those in our case and save some CPU, as we don't have a need to lint those non-source-code files.
+            return nil // just skip those (but don't throw — those are valid use cases in any pbxproj, just ones we don't care about)
```
FYI:

In a previous commit (8bc174acbc6208b78f28f8aa058a4ddcf45fb9cf) I started to handle those cases of `PBXBuildFile`s pointing to a `PBXVariantGroup` (e.g. `Localizable.strings`) or `XCVersionGroup` (e.g. `Model.xcdatamodeld`) and for those cases I'd just (1) iterate over each of their `children` to resolve _their_ URL (2) and then make `resolveURL` ultimately return an array of `URL`, (because now a single `PBXBuildFile` could represent the building of a _set_ of related files, not always a single one).

But ultimately (44dd3ed99ca0ece357e53cbb5cf74132de016553) I figured it was not worth the effort because the only cases where this happen are for localized resources (`PBXVariantGroup`) and versioned resources (`XCVersionGroup`)… which source files like `.swift` and `.m` never are. And since we're only interested in those source files and not in resources for that linter… better keep it simple after all 🙃

AliSoftware replied:
So TL;DR: now we only try to resolve URLs for `PBXBuildFiles` which:
1. Have a `fileRef` value (ignoring the ones with `productRef`), and
2. Point to a `PBXFileReference` (ignoring the ones pointing to a `PBXVariantGroup` or `XCVersionGroup`)

AliSoftware commented on Scripts/BuildPhases/LintAppLocalizedStringsUsage.swift:
```diff
@@ -95,18 +102,25 @@ protocol PBXReference: PBXObject {
 extension Xcodeproj {
     /// An error `thrown` when an inconsistency is found while parsing the `.pbxproj` file.
-    enum DecodingError: Swift.Error, CustomStringConvertible {
+    enum ProjectInconsistencyError: Swift.Error, CustomStringConvertible {
```
That made me wonder about naming: should we call such errors `InconsistencyError` (as in "it is an error about an inconsistency") or… `ConsistencyError` (as in "it is an error in the consistency of the file")? 🤔 😅 

Eh. 🤷

mokagio replied:
"There's an error with the consistency, so it's inconsistent, which means there's an error with its consistency...

![image](https://user-images.githubusercontent.com/1218433/169218452-9fea7ac8-5b5e-4ca9-836e-f0f6b6a0e499.png)

AliSoftware commented on Scripts/BuildPhases/LintAppLocalizedStringsUsage.swift:
```diff
@@ -115,135 +129,115 @@ extension Xcodeproj {
             return sourceBuildPhase.files.compactMap(object(id:)) as [PBXBuildFile]
         }
     }
-
-    /// Helper type to ensure the `isa` field of a `PBXObject` contains the name of the expected type to decode as its value.
-    struct ISA<T>: RawRepresentable, Decodable, CustomDebugStringConvertible {
-        var rawValue: String
-        init?(rawValue: String) {
-            guard rawValue == String(describing: T.self) else { return nil }
-            self.rawValue = rawValue
-        }
-        var debugDescription: String { self.rawValue }
-    }
-
```
Bye bye ingenious but slightly convoluted solution 👋 😅

AliSoftware commented on Scripts/BuildPhases/LintAppLocalizedStringsUsage.swift:
```diff
@@ -115,135 +129,115 @@ extension Xcodeproj {
         let buildPhases: [ObjectUUID]
-        let productType: ProductType
-
+        let productType: String
+        var knownProductType: ProductType? { ProductType(rawValue: productType) }
+        
```
This is to allow potential unknown `productType` values in future versions of Xcode.
We now allow _any_ `String` value (instead of failing to parse the project's plist if we encounter an unknown `case`), and then `knownProductType` still exposes the nice typed value… but only if known.

AliSoftware replied:
To be honest initially I implemented this case using a Property Wrapper and I was happy with myself because it looked so beautiful… I mean look at this magic: eacf33e5c3237a062cccda06cce37d316fcf868b 🤩 

But ultimately I accepted that simpler was probably better… and said goodbye to my beloved Property Wrapper in 70f963075762f89fc2fde23d1550cc7a9ca7dde3 after all 😅

mokagio replied:
Well done. I know it mustn't have been easy. 💪

AliSoftware commented on Scripts/BuildPhases/LintAppLocalizedStringsUsage.swift:
```diff
@@ -115,135 +129,115 @@ extension Xcodeproj {
         case buildProductsDir = "BUILT_PRODUCTS_DIR"
         case devDir = "DEVELOPER_DIR"
         case sdkDir = "SDKROOT"
+        var description: String { rawValue }
```
For nicer and less verbose error messages (otherwise the full qualified name of the enum was printed in `DecodingError`s thrown by the `PropertyListDecoder` on invalid `pbxproj` file decoding)

AliSoftware commented on Scripts/BuildPhases/LintAppLocalizedStringsUsage.swift:
```diff
@@ -115,135 +129,115 @@ extension Xcodeproj {
     /// One of the many `PBXObject` types encountered in the `.pbxproj` file format.
     /// Represents a group (aka "folder") contained in the project tree.
     struct PBXGroup: PBXReference {
-        enum ISA: String, Decodable { case PBXGroup, XCVersionGroup, PBXVariantGroup }
-        // We don't have a `ISA<Self>` here because we want multiple `isa` values to all be allowed and all decode as a `PBXGroup` instance (`"PBXGroup"`, `"XCVersionGroup"`, `"PBXVariantGroup"`)
-        private let isa: ISA
-
```
No more! 👋

AliSoftware commented on Scripts/BuildPhases/LintAppLocalizedStringsUsage.swift:
```diff
@@ -252,16 +246,14 @@ extension Xcodeproj {
-                    self.wrappedValue = object
-                    return
-                }
+            let untypedObject = try UnknownPBXObject(from: decoder)
+            if let objectType = Self.knownTypes.first(where: { $0.isa == untypedObject.isa }) {
+                self.wrappedValue = try objectType.init(from: decoder)
+            } else {
+                self.wrappedValue = untypedObject
             }
-            self.wrappedValue = try UnknownPBXObject(from: decoder) as PBXObject // Fallback
         }
```
The new trick (instead of the `ISA<Self>` we had previously) is simply to call `init(from:)` on 2 types (`UnknownPBXObject` then the dynamic `objectType` type.

Each call to `init(from:)` at that level will each request a `KeyedDecodingContainer` on the same "dictionary" found at the same level in the plist (it's just that each type will be using `decoder.container(keyedBy:)` with different `CodingKeys` for each).

Which wouldn't be achievable if we implemented this `PBXObjectWrapper.init(from decoder: Decoder)` using the classical pattern of calling `decoder.container(keyedBy: ISAKey.self)` then `container.decode(objectType, forKey: ???)` to try to decode the rest of the fields (and is what made all this "decode object type based on one its field" a challenge to begin with)

---

Some benefits of this approach compared to the old one:
 - It should be more performant, as it decodes the `isa` field as a `String` (under the hood via `UnknownPBXObject(from: decoder)`, then immediately known which `objectType` to try to decode from there, as opposed to having to try to decode all the known types and let many fail until we find one that doesn't, like we did before
 - It also allows better error handling, because if we have a case of `isa = "PBXFileReference"` but for some reason (e.g. corrupted `pbxproj` file) the other fields were invalid and not matching an actual `PBXFileReference` making `PBXFileReference(from: decoder)` fail to decode, then:
    - The previous approach would just silently eat that error and fallback to a `UnknownPBXObject` instead — despite us knowing about what the `PBXFileReference` is supposed to be and wanting our script to handle them — and would not report the inconsistency and the project being corrupted
    - While with that new approach, if the `isa` tells us that it should be a `PBXFileReference` but we fail to decode it as such, we can `throw` a proper corresponding error instead of silently ignoring the inconsistency.

And obviously one last benefit is that it's much simpler… but it just took me some time to figure out that this would be the working solution (as it's not following the standard pattern for `init(from:)` implementations and is a bit unconventional)

AliSoftware commented on Scripts/BuildPhases/LintAppLocalizedStringsUsage.swift:
```diff
@@ -39,54 +39,61 @@ class Xcodeproj {
         switch object.sourceTree {
         case .absolute:
-            return URL(fileURLWithPath: object.path!)
+            guard let path = object.path else { throw ProjectInconsistencyError.incorrectAbsolutePath(id: objectUUID) }
+            return URL(fileURLWithPath: path)
```
[Save those ponies](https://alisoftware.github.io/swift/2015/09/06/thinking-in-swift-1/)! 😄 🐴

mokagio commented:
I haven't had time to review this in time for the 19.9 code freeze. Given it's not a user facing PR, I'm bumping it to 20.0 to prioritize getting started with the 19.9 code freeze.

mokagio commented on Scripts/BuildPhases/LintAppLocalizedStringsUsage.sh:
```diff
@@ -8,10 +8,14 @@ LINTER_EXEC="${LINTER_BUILD_DIR}/$(basename "${SCRIPT_SRC}" .swift)"
 
-"$LINTER_EXEC" "${PROJECT_FILE_PATH:-$1}" # "${TARGET_NAME:-$2}"
+if [ -z "${PROJECT_FILE_PATH:=${1:-}}" ]; then
+  echo "Please provide the path to the xcodeproj to scan"
```
What do you think of adding the "error:" prefix, which should make Xcode report this in the navigator?

```suggestion
  echo "error: Please provide the path to the xcodeproj to scan"
```

AliSoftware replied:
Applied in [eb3cf3e](https://github.com/wordpress-mobile/WordPress-iOS/pull/18525/commits/eb3cf3effc1f1a1c075ea791422b90555e17336a)

mokagio commented on Scripts/BuildPhases/LintAppLocalizedStringsUsage.swift:
```diff
@@ -300,26 +292,34 @@ func lint(fileAt url: URL, target: String) throws -> LintResult {
+    print("Done! \(violationsFound) violation(s) found.")
+    exit(violationsFound > 0 ? 1 : 0)
+} catch let error {
+    print("\n\nError while parsing `\(projectPath)`: \(error)")
```
What do you think of adding the "error:" prefix, which should make Xcode report this in the navigator?

```suggestion
    print("\n\nerror: Error while parsing `\(projectPath)`: \(error)")
```

<img width="608" alt="image" src="https://user-images.githubusercontent.com/1218433/169220819-872ac94a-fc8e-427f-8b2f-7e9d8c1d1c0b.png">

AliSoftware replied:
Actually we should also be able to include the path as well in the part interpreted by Xcode I think:

```suggestion
    print("\n\n\(projectPath): error: \(error)")
```

Let me try by intentionally create an error and see.

(Loved your 🥔 object, btw 😂 )

AliSoftware replied:
Applied in [eb3cf3e](https://github.com/wordpress-mobile/WordPress-iOS/pull/18525/commits/eb3cf3effc1f1a1c075ea791422b90555e17336a)

PS: Note that in my experiments, Xcode does not always report the projectPath in the main "Error" UI, hence why I kept it in the error message as well to make sure it would always be visible regardless of context and Xcode being Xcode.

mokagio reviewed (approved):
Smoke tested locally and worked well, just like before 👍

AliSoftware committed:
Apply error reporting wording suggestions

https://github.com/wordpress-mobile/WordPress-iOS/pull/18525#discussion_r876622089
https://github.com/wordpress-mobile/WordPress-iOS/pull/18525#discussion_r876616787

Co-authored-by: Gio Lodi <giovanni.lodi42@gmail.com>
